Tasks/SonarQubePostTest/SonarQubeHelper.ps1

```diff
@@ -0,0 +1,230 @@
+
+
+# When passing arguments to a process, the quotes need to be doubled and   
+# the entire string needs to be placed inside quotes to avoid issues with spaces  
+function EscapeArg  
+{  
+    param([string]$argVal)  
+  
+    $argVal = $argVal.Replace('"', '""');  
+    $argVal = '"' + $argVal + '"';  
+  
+    return $argVal;  
+}  
+
+
+# Set a variable in a property bag that is accessible by all steps
+# To retrieve the variable use $val = Get-Variable $distributedTaskContext "varName"
+function SetTaskContextVariable
+{
+    param([string][ValidateNotNullOrEmpty()]$varName, 
+          [string]$varValue)
+    
+    Write-Host "##vso[task.setvariable variable=$varName;]$varValue"
+}
+
+function GetTaskContextVariable()
+{
+    param([string][ValidateNotNullOrEmpty()]$varName)
+    return Get-TaskVariable -Context $distributedTaskContext -Name $varName
+}
+
+#
+# Helper that informs if a "filePath" has been specified. The platform will return the root of the repo / workspace if the user enters nothing.
+#
+function IsFilePathSpecified
+{
+     param([string]$path)
+
+     if ([String]::IsNullOrWhiteSpace($path))
+     {
+        return $false
+     }
+
+     return ![String]::Equals(
+                [System.IO.Path]::GetFullPath($path).TrimEnd([System.IO.Path]::DirectorySeparatorChar, [System.IO.Path]::AltDirectorySeparatorChar),
+                [System.IO.Path]::GetFullPath($env:BUILD_SOURCESDIRECTORY).TrimEnd([System.IO.Path]::DirectorySeparatorChar, [System.IO.Path]::AltDirectorySeparatorChar),
+                [StringComparison]::OrdinalIgnoreCase)
+}
+
+
+#
+# Retry a specific command until it completes without exceptions. Usage:  
+#
+#   $command = { InvokeGetRestMethod "/api/projects"} 
+#   Retry $command -maxRetries 30 -retryDelay 1 -Verbose
+#
+function Retry
+{
+    [CmdletBinding()]
+    param (    
+    [Parameter(ValueFromPipeline,Mandatory)]$command,
+    [int]$maxRetries = 3, 
+    [int]$retryDelay = 1)    
+
+    $success = $false
+    $attemptNumber = 1
+
+    while (!$success)
+    {
+        try
+        {
+            Write-Verbose "Calling $command"
+            $result = & $command
+            $success = $true
+        }
+        catch
+        {
+            if ($attemptNumber -ige $maxRetries) 
+            {
+                Write-Verbose "Calling $command failed after $attemptNumber retries"
+                throw
+            }
+            else
+            {
+                Write-Verbose "Calling $command failed. Attempt number $attemptNumber. Retrying after $retryDelay seconds..."
+                Start-Sleep $retryDelay
+                $attemptNumber++
+            }
+        }
+    }
+
+    return $result
+}
+
+#
+# Retry a specific command until the results is true. Returns true if the command succeded. Usage:  
+#
+#   $command = { GetStatus -eq "Complete" } 
+#   RetryUntilTrue $command -maxRetries 30 -retryDelay 1 -Verbose
+# 
+#
+function RetryUntilTrue
```

Comment:
Retry invokes the command until no exceptions are thrown while RetryUntilTrue invokes the command until it returns true. I'm not sure that having one use the other will simplify implementation, especially since PS is not well equipped to deal with delegates